state/execution.go

```diff
@@ -281,6 +283,12 @@ func execBlockOnProxyApp(
 		return nil, err
 	}
 
+	if len(abciResponses.EndBlock.Tags) > 0 {
+		tag := abciResponses.EndBlock.Tags[len(abciResponses.EndBlock.Tags)-1]
+		if bytes.Equal(tag.Key, []byte(UpgradeTagKey)) {
+			return nil, fmt.Errorf(string(tag.Value))
+		}
```

Comment:
How about go though all tags and find the 'UpgradeTagKey' instead of the tricky 'last one element' ?